src/rt-app_parse_config.c

```diff
@@ -983,37 +983,94 @@ parse_task_phase_data(struct json_object *obj,
 {
 	/* used in the foreach macro */
 	struct lh_entry *entry; char *key; struct json_object *val; int idx;
-	int i;
+	struct json_object *array_val;
+	size_t i;
+	struct json_object *events_array = NULL;
+	bool is_first_entry;
 
 	log_info(PFX "Parsing phase");
 
+	data->nbevents = 0;
+	data->events = NULL;
+
 	/* loop */
 	data->loop = get_int_value_from(obj, "loop", TRUE, 1);
 
-	/* Count number of events */
-	data->nbevents = 0;
 	foreach(obj, entry, key, val, idx) {
-		if (obj_is_event(key))
+		if(!strncmp("events", key, 6)) {
+			if(!json_object_is_type(val, json_type_array)) {
+				log_critical(PIN "\"events\" key must be an array");
+				exit(EXIT_INV_CONFIG);
+			}
+			events_array = val;
+			break;
+		}
+	}
+
+	foreach(obj, entry, key, val, idx) {
+		if (obj_is_event(key)) {
+			/* Check that we are not trying to mix old and new styles */
+			if (events_array) {
+				log_critical(PIN "Event key %s cannot be used in conjunction with \"events\" key", key);
+				exit(EXIT_INV_CONFIG);
+			/* Count number of events */
+			} else {
 				data->nbevents++;
+			}
+		}
 	}
 
+	if (events_array)
+		/* Assumes that all item in the array will be valid. If an item
+		 * does not yield an entry in data->events, we must ensure it won't
+		 * be used by exiting
+		 */
+		data->nbevents = json_object_array_length(events_array);
+
 	if (data->nbevents == 0) {
 		log_critical(PIN "No events found. Task must have events or it's useless");
 		exit(EXIT_INV_CONFIG);
-
 	}
 
 	log_info(PIN "Found %d events", data->nbevents);
-
 	data->events = malloc(data->nbevents * sizeof(event_data_t));
 
 	/* Parse events */
-	i = 0;
-	foreach(obj, entry, key, val, idx) {
-		if (obj_is_event(key)) {
-			log_info(PIN "Parsing event %s", key);
-			parse_task_event_data(key, val, &data->events[i], tdata, opts);
-			i++;
+	if (events_array) {
+		for (i=0; i < data->nbevents; i++) {
+			array_val = json_object_array_get_idx(events_array, i);
+			is_first_entry = true;
+			foreach(array_val, entry, key, val, idx) {
```

Comment:
what is the purpose of this foreach ?
isn't all object inside the array an event

Reply:
The array items are themselves objects:
```
[{event1: param1}, {event2: param2}, ...]
```

The alternative would be to use sub-arrays like a tuple instead of objects, but the object approach is the standard way of doing things, as it is pretty clean in YAML:

```
events:
    - run: 42
    - sleep: 53
 ```

(and the sub-array solution does not save any typing anyway)

Reply:
but that was already the case with the old style and obj_is_event() is there to check that those objects are events

Reply:
After checking parse_task_event_data() it looks like it's doing a validation on the event name and will exit if it's given a non-event so we can skip this check here.

We still need to check the number of keys though, as:
* 0 keys would mean that the length of the array (and therefore the size of data->events) does not match the actual number of events

* more than one key does not make sense (at least for now, maybe one day this could be used to attach some sort of metadata to the event)

I need to fix the code to check we are not in the 0 keys situation, currently it would trigger a UB with `data->events[i]` not being initialized.